src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -57,59 +58,73 @@ public override int Read(Span<byte> buffer)
+                }
 
-                    if (_bufferCount <= 0)
+                // If we successfully decompressed any bytes, or if we've reached the end of the decompression, we're done.
+                if (bytesWritten != 0 || lastResult == OperationStatus.Done)
+                {
+                    break;
+                }
+
+                if (buffer.IsEmpty)
+                {
+                    // The caller provided a zero-byte buffer.  This is typically done in order to avoid allocating/renting
+                    // a buffer until data is known to be available.  We don't have perfect knowledge here, as _decoder.Decompress
+                    // will return DestinationTooSmall whether or not more data is required.  As such, we assume that if there's
+                    // any data in our input buffer, it would have been decompressible into at least one byte of output, and
+                    // otherwise we need to do a read on the underlying stream.  This isn't perfect, because having input data
+                    // doesn't necessarily mean it'll decompress into at least one byte of output, but it's a reasonable approximation
+                    // for the 99% case.  If it's wrong, it just means that a caller using zero-byte reads as a way to delay
+                    // getting a buffer to use for a subsequent call may end up getting one earlier than otherwise preferred.
+                    Debug.Assert(lastResult == OperationStatus.DestinationTooSmall);
+                    if (_bufferCount != 0)
                     {
                         break;
                     }
                 }
 
-                lastResult = _decoder.Decompress(new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount), buffer, out int bytesConsumed, out int bytesWritten);
-                if (lastResult == OperationStatus.InvalidData)
+                // Read more data from the underlying stream.
+                Debug.Assert(
+                    lastResult == OperationStatus.NeedMoreData ||
+                    (lastResult == OperationStatus.DestinationTooSmall && _bufferCount == 0), $"{nameof(lastResult)} == {lastResult}, {nameof(buffer.Length)} == {buffer.Length}");
+
+                // Ensure any left over data is at the beginning of the array so we can fill the remainder.
+                if (_bufferCount != 0 && _bufferOffset != 0)
                 {
-                    throw new InvalidOperationException(SR.BrotliStream_Decompress_InvalidData);
+                    new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount).CopyTo(_buffer);
                 }
+                _bufferOffset = 0;
 
-                if (bytesConsumed > 0)
+                int numRead = _stream.Read(_buffer, _bufferCount, _buffer.Length - _bufferCount);
+                if (numRead <= 0)
                 {
-                    _bufferOffset += bytesConsumed;
-                    _bufferCount -= bytesConsumed;
+                    break;
```

Comment:
It feels like this is missing a case where the underlying stream returns EOF but the decompressor is still expecting more data. That should be an error, right?

Reply:
Maybe?  Both DeflateStream and BrotliStream behave this way already.  I'm not sure if changing it to throw is the right thing to do (I'm also not actually sure how to query the decoder in this case for that state... would need to look into it).  I'll leave it as-is for now, but we can open an issue if you think it's worth pursuing.